New runtime & gamestate

gecotron opened the pull request:
Based on previous work & discussions. Please let me know any tweaks & changes needed

gecotron committed:
feat: basic callstack

My own implementation of a callstack. Missing drawstack rn but thats not that hard to add

gecotron committed:
feat: basic callstack implementation

quick test of the callstack on main. Not really there yet in terms of the overall vision, but I think it does the job alright so far

gecotron committed:
chore: ide spellcheck fix

callstack is a word. At least to me

gecotron committed:
chore: workflows!

bored and want workflows for this branch

gecotron committed:
fix: proper use of callstacks now yipee

Fixes up implementation of traits for Call items. They now work all nice and proper

gecotron committed:
fix: proper implementation of Scene/logic stack

Still barebones, but doing the job

gecotron committed:
fix: movement more grid-like

Moves like on a grid now, but still not on a grid... More to come

gecotron committed:
chore: using (haha) prelude properly

Values properly exposed now

gecotron committed:
feat: default-ish implementations for references of trait objects for  trait

Avoids writing code for references & mutable references of any struct that implements the Call trait. Less code down the line

gecotron committed:
fix: using Call trait properly

Minor fixups to implementing the Call trait. Nothing else really

gecotron committed:
chore: fixup prelude

See commit 8ad28da1ba8ebbbd8fa9c4d6cef952b46fafa7b4

gecotron committed:
chore: ide spellcheck again

I hate this part so much

gecotron committed:
feat!: restructure runtime functions and related data

Based on issue #17, all runtime related data has been moved to the main file. Aside from that there are a lot more comments now

gecotron committed:
feat: comments!

words for your eyes

phynxinferno commented on .github/workflows/rust-windows.yml:
```diff
@@ -2,7 +2,7 @@ name: Rust Windows
 
 on:
   push:
-    branches: ["main"]
```
Maybe change this back on merge

gecotron replied:
:+1:

phynxinferno commented:
Looks good so far. Let's avoid merging for now, I'll test some changes at home.

phynxinferno commented on src/main.rs:
```diff
@@ -94,3 +70,62 @@ fn window_conf() -> Conf {
+    fn push_mut<T: Call>(&mut self, item: &'s mut T) {
+        self.function_stack.push(Box::new(item));
+    }
+
```
Looking good so far but we need to make the drawstack calls more ambiguous

gecotron replied:
Will do. Function pointer passing is a thing now with more recent commits on the branch, if we can, I think updating/making a new pull request will probably be in order soon. Not immediately though, there is still some work to be done

gecotron committed:
feat: we documenting it with this one

doc files!

gecotron committed:
feat: documentation action

We documenting the project with this one

gecotron committed:
style: relocate files to be more in line with current rust formatting styles

gecotron committed:
feat: Levels, maps & loading TOML config files

We approach an actual game you can play every single day. It is wonderful. I have implemented a skeleton (as usual) of a level and map system to play on. Level configs can be stored in a toml file.

gecotron commented:
Quick note. This probably should not be merged, as the mdw-gamestate branch has more updates to the system, and a myriad of other work there that should be all dumped in one go. Another pr will probably be made

gecotron committed:
fix: Miku under a hydraulic press

resized & compressed to 100x100 pixels

gecotron committed:
feat: refactoring & grids

refactored the initialization of player and the test level. Added a function to draw a grid onscreen to debug and waste memory

gecotron committed:
feat: check player's position related to tiles. Useful later on for actual gaming. Player can now "jump" 2 tiles, and is properly & nicely rendered on screen

gecotron committed:
feat: Rendering the MAP

the cubes represent all.

gecotron committed:
feat!: Major refactor of the gamestate & runtime features

The gamestate is now ONLY gamestate. A separate stack for non-persistent Call entities (mainly levels & menus) that are actively triggered on events. Be aware there are breaking changes

gecotron committed:
fix: pure grid movement

player moves only on the grid, and the grid updates its position! Ingenious!

gecotron committed:
feat: including gamefiles in binary

This is just mainly a test for the future. However, asset files can and now should be embedded in the binary for the final result

gecotron committed:
chore: miku miku BEEAAM!!

spellcheck doesn't believe miku exists for some reason

gecotron committed:
fix: move tiles to a more understandable spot

tiles now sit with the map. made more sense

gecotron committed:
feat: overhaul call objects

In progress, but now return runcodes! working on making them actually functional, because they arent right now

gecotron committed:
chore: update test map

added win tile. to be used to test final lap, winning the game

gecotron committed:
feat: full win lose implementation

This game is like 80-90% complete. What do you know

gecotron committed:
feat: can you feel the rythm?

probably not. but you will soon!

gecotron commented:
Closed due to language switch
